tests/test_client.py

```diff
@@ -113,115 +174,214 @@ def metadata_client():
+                client_stream_input
+            ).output
+        else:
+            client_stream_output = await client.ExecuteClientStream(
+                client_stream_input
+            )
+            client_stream_output = client_stream_output.output
     assert metadata_string in client_stream_output
 
 
-def test_metadata_client_server_stream(metadata_client, metadata_string):
+@pytest.mark.parametrize("aio", [False, True])
+async def test_metadata_client_server_stream(metadata_string, aio):
     """Invocation metadata should be added to the servicer context."""
-    stream_stream_input = iter((DummyRequest(input="metadata"),))
-    result = metadata_client.ExecuteClientServerStream(stream_stream_input)
-    stream_stream_output = [r.output for r in result]
+    intr = AsyncMetadataInterceptor([("this_key", "this_value")]) if aio else MetadataInterceptor(
+        [("this_key", "this_value")])
+    interceptors = [intr]
+
+    special_cases = {
+        "metadata": lambda _, c: ",".join(
+            f"{key}:{value}" for key, value in c.invocation_metadata()
+        )
+    }
+    with dummy_client(
+            special_cases=special_cases, client_interceptors=interceptors, aio_server=aio, aio_client=aio,
+    ) as client:
+        stream_stream_input = iter((DummyRequest(input="metadata"),))
+        if not aio:
+            result = client.ExecuteClientServerStream(stream_stream_input)
+            stream_stream_output = [r.output for r in result]
+        else:
+            result = client.ExecuteClientServerStream(stream_stream_input)
+            stream_stream_output = [r.output async for r in result]
     assert metadata_string in "".join(stream_stream_output)
 
 
-def test_code_counting():
+@pytest.mark.parametrize("aio", [False, True])
+async def test_code_counting(aio):
     """Access to code on call details works correctly."""
-    interceptor = CodeCountInterceptor()
+    interceptor = AsyncCodeCountInterceptor() if aio else CodeCountInterceptor()
     special_cases = {"error": raises(ValueError("oops"))}
     with dummy_client(
-        special_cases=special_cases, client_interceptors=[interceptor]
+        special_cases=special_cases, client_interceptors=[interceptor], aio_server=aio, aio_client=aio,
     ) as client:
         assert interceptor.counts == {}
-        client.Execute(DummyRequest(input="foo"))
-        assert interceptor.counts == {grpc.StatusCode.OK: 1}
-        with pytest.raises(grpc.RpcError):
-            client.Execute(DummyRequest(input="error"))
-        assert interceptor.counts == {grpc.StatusCode.OK: 1, grpc.StatusCode.UNKNOWN: 1}
+        if not aio:
+            client.Execute(DummyRequest(input="foo"))
+            assert interceptor.counts == {grpc.StatusCode.OK: 1}
+            with pytest.raises(grpc.RpcError):
+                client.Execute(DummyRequest(input="error"))
+            assert interceptor.counts == {grpc.StatusCode.OK: 1, grpc.StatusCode.UNKNOWN: 1}
```

Comment:
Yank this out of the aio conditional? because it should be the same on either branch.

Nit: I find it to be a lot less error prone to test the positive case rather than the negative case, i.e. `if aio` vs `if not aio`, it just helps lead to fewer double and triple negatives.